faithcomesbyhearing/dbp
# FCBHDBP-188 hardening: optimize slow query - insert into project_member

## hardening: optimize slow query - insert into project_member

# Description
The main query to know if user has a project_members has been updated to use the from_date filter. The syncV2Users command use the `from_date` value to filter the user pulled from `dbp_users_v2` and it inserts into dbp_users. Then, we have used this filter to know if user that belongs to dbp_users has a project_member record. Also, I have improved the endpoint to pull the users because this endpoint was using the same query that the syncV2Users command.

- Old query:

```sql
SELECT 52341 AS project_id, id AS user_id, 1 AS role_id, u.created_at
FROM users AS u
LEFT JOIN project_members AS pm ON pm.user_id = u.id
where u.v2_id > 0 AND pm.project_id IS NULL;
```

- New query:
```sql
SELECT 52341 AS project_id,
id AS user_id,
1 AS role_id,
u.created_at
FROM `dbp_users`.users AS u 
where created_at >= '2019-08-20 12:08:00' AND u.v2_id > 0 NOT IN (
    SELECT `dbp_users`.`project_members`.`user_id` FROM `dbp_users`.`project_members`
);
```

## Issue Link
https://fullstacklabs.atlassian.net/browse/FCBHDBP-188

## How Do I QA This
you should execute the command:

```bash
php artisan syncV2:users
```

The query to insert the rows `project_members` should execute in less time than 17 seconds. I have executed on my local environment and it has taken around 3 seconds.
